polygon/heimdall/version_monitor.go

```diff
@@ -0,0 +1,100 @@
+package heimdall
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/erigontech/erigon-lib/log/v3"
+)
+
+type HeimdallVersion int64
+
+const (
+	HeimdallV1 HeimdallVersion = iota
+	HeimdallV2
+)
+
+type heimdallClient interface {
+	FetchStatus(ctx context.Context) (*Status, error)
+}
+
+type versionMonitor struct {
+	ctx            context.Context
+	currentVersion HeimdallVersion
+
+	heimdall heimdallClient
+	logger   log.Logger
+
+	frequency time.Duration
+	m         sync.Mutex
+	once      sync.Once
+}
+
+func NewVersionMonitor(ctx context.Context, heimdallClient heimdallClient, logger log.Logger, frequency time.Duration) *versionMonitor {
+	return &versionMonitor{
+		ctx:            ctx,
+		heimdall:       heimdallClient,
+		currentVersion: HeimdallV1,
+		frequency:      frequency,
+	}
+}
+
+func (vm *versionMonitor) Run() {
+	checkEvery := time.NewTicker(vm.frequency)
+	defer checkEvery.Stop()
+
+	for {
```

Comment:
Not quite sure why we need this, is this due to some odd Heimdall behaviour? We don't expect the version to change between runs, perhaps this code could be simplified.

Reply:
I had this conversation with Polygon team and they confirmed this is a valid scenario, this is why I added a couple of unit tests to cover it: TestVersioMonitorHeimdallUpgrade and TestVersioMonitorHeimdallDowngrade